lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp

```diff
@@ -655,6 +665,87 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+    ConversionPatternRewriter &rewriter) const {
+  MLIRContext *ctx = op.getContext();
+  Location loc = op.getLoc();
+  StringAttr kRegister = str_attr("register");
+  StringAttr kLane = str_attr("lane");
+  assert(!cvtNeedsSharedMemory(op.getSrc().getType(), op.getType()));
+  auto [P1, Cp, P2inv] = std::move(decomposed);
+
+  // Grab the source elements and prepare the outputs of just the shuffles.
+  SmallVector<Value> srcValues =
+      unpackLLElements(loc, adaptor.getSrc(), rewriter);
+  SmallVector<Value> shflOuts(Cp.getInDimSize(kRegister));
+
+  Value threadId = getThreadId(rewriter, loc);
+  Value threadsPerWarp = i32_val(Cp.getInDimSize(kLane));
+  Value laneId = urem(threadId, threadsPerWarp);
+
+  // To minimize the number of selects emitted on the source side, determine the
+  // minimum set of registers that could be selected from each thread.
+  // InstCombine *might* be able to crush this, but if the sizePerThread is
+  // large, it's truly a huge number of selects that get emitted.
+  // If reducedP1 is trivial, then we will emit
+  // shflSrc = select(i == i, src[i], undef) and this will get trivially folded,
+  // so don't worry about this case.
+  LinearLayout reducedP1 = P1.removeZeroBasesAlongDim(kLane);
+
+  // Emit one shuffle per destination register.
+  for (int i = 0, e = shflOuts.size(); i != e; ++i) {
+    // 'Cp' maps a (dst_lane, dst_reg) -> (src_lane, src_reg), and we know that
+    // for a register, it does not map to different registers in the same lane.
+    // At the same time, for each register, P1 returns the source value index
+    // to provide as the shuffle value.
+    auto out = applyLinearLayout(loc, rewriter, P1,
+                                 {{kLane, laneId}, {kRegister, i32_val(i)}});
+    assert(out.size() == 1);
+    Value srcRegIdx = out.front().second;
+    // The size of the input lane dimension is the number of selects to emit.
+    // TODO(jeff): For dtypes smaller than i32, we can use byte permutes and
+    // shuffle multiple values at a time.
+    Value shflSrc = undef(srcValues.front().getType());
+    for (unsigned j = 0, e = reducedP1.getInDimSize(kLane); j != e; ++j) {
+      int32_t check =
+          reducedP1.apply({{kLane, j}, {kRegister, i}}).front().second;
+      shflSrc =
+          select(icmp_eq(srcRegIdx, i32_val(check)), srcValues[check], shflSrc);
+    }
+
+    out = applyLinearLayout(loc, rewriter, Cp,
+                            {{kLane, laneId}, {kRegister, i32_val(i)}});
+    assert(out.size() == 1);
+    Value shflIdx = out.front().second;
+    shflOuts[i] = targetInfo.shuffleIdx(rewriter, loc, shflSrc, shflIdx);
+  }
+
+  // Finally, we just need to apply P2 to the shflOuts to permute the registers
+  // into their final form. Use the same trick to reduce the number of emitted
+  // selects.
+  LinearLayout reducedP2 = P2inv.removeZeroBasesAlongDim(kLane);
+  SmallVector<Value> results(shflOuts.size());
+  for (int i = 0, e = results.size(); i != e; ++i) {
```

Comment:
why this format??

Reply:
Old habits die hard. This is from the upstream MLIR coding style guides. The idea is that if you re-query the size in the condition, it's indication that you expect the size to change, e.g.

```C++
for (auto i = 0; i < astNodes.size(); ++i) {
  processMoreNodesRecursively(astNodes[i], &astNodes);
}
```

I'm not really sure what our coding style is other than a mishmash of the styles of other MLIR codebases I've seen lol